zcash_primitives/src/sapling/zip32.rs

```diff
@@ -231,12 +229,34 @@ impl DiversifierKey {
     }
 }
 
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum KeyIndex {
+    Master,
+    Child(ChildIndex),
+}
+
+impl KeyIndex {
+    fn new(depth: u8, i: u32) -> Option<Self> {
+        match i {
+            0 if depth == 0 => Some(KeyIndex::Master),
+            _ => ChildIndex::from_index(i).map(KeyIndex::Child),
+        }
+    }
+
+    fn index(&self) -> u32 {
```

Comment:
Document that this index has the $2^{31}$ hardened bit set. (ZIP 32, following BIP 32, uses a poorly chosen notation that obscures whether that bit is part of the index or not. If you read carefully it is for child indices, but not for account indices, so it's easy to get confused.)

Reply:
In this PR, I clean up the types and methods to consistently use "index" to mean the ZIP 32 notion, and "value" to mean the 31-bit value without the hardening bit.